src/main/java/org/broadinstitute/hellbender/engine/filters/GoodCigarFilter.java

```diff
@@ -0,0 +1,83 @@
+package org.broadinstitute.hellbender.engine.filters;
+
+import htsjdk.samtools.Cigar;
+import htsjdk.samtools.CigarOperator;
+import htsjdk.samtools.SAMRecord;
+
+import java.util.regex.Pattern;
+
+/**
+ * Filter out reads with wonky cigar strings.
+ *
+ *  - No reads with a different length and cigar length
+ *  - No reads with Hard/Soft clips in the middle of the cigar
+ *  - No reads starting with deletions (with or without preceding clips)
+ *  - No reads ending in deletions (with or without follow-up clips)
+ *  - No reads that are fully hard or soft clipped
+ *  - No reads that have consecutive indels in the cigar (II, DD, ID or DI)
+ *
+ * From SAM spec:
+ *  - H can only be present as the first and/or last operation
+ *  - S may only have H operations between them and the ends of the CIGAR string
+ *
+ *  ps: apparently an empty cigar is valid.
+ */
+public final class GoodCigarFilter implements ReadFilter {
+
+    private static final String DIGITS = "\\d*"; //Any operator can be preceeded by a series of digits.
+    private static final String M = cat(DIGITS, CigarOperator.M.name());
+    private static final String H = cat(DIGITS, CigarOperator.H.name());
+    private static final String S = cat(DIGITS, CigarOperator.S.name());
+    private static final String I = cat(DIGITS, CigarOperator.I.name());
+    private static final String D = cat(DIGITS, CigarOperator.D.name());
+    private static final String N = cat(DIGITS, CigarOperator.N.name());
+    private static final String P = cat(DIGITS, CigarOperator.P.name());
+    private static final String EQ= cat(DIGITS, CigarOperator.EQ.name());
+    private static final String X = cat(DIGITS, CigarOperator.X.name());
+
+    private static final String NOT_H    = any(M, S, I, D, N, P, EQ, X);
+    private static final String ANYTHING = maybeSome(any(NOT_H, H));
+    private static final String CLIP     = any(H, S);
+
+    //We pre-create one regular expression pattern and reuse it.
+    //Each clause in the pattern corresponds to one condition we're testing for.
+    private final static Pattern BAR_CIGAR = Pattern.compile(
+           any((ANYTHING + NOT_H + H + NOT_H + ANYTHING),                      //H in the middle without flanking H
+               (ANYTHING + NOT_H + S + NOT_H + ANYTHING),                      //S in the middle without flanking H
+               (ANYTHING + any(I+I, D+D, I+D, D+I) + ANYTHING),                //consecutive indels
+               (maybeSome(CLIP)+ D + ANYTHING),                                //starts with deletion
+               (ANYTHING + D + maybeSome(CLIP)),                               //ends with deletion
+               (some(CLIP))));                                                 //only clips
```

Comment:
This approach makes me incredibly nervous -- complex regular expressions are notoriously prone to edge conditions and hard-to-see bugs, especially when (as here) there are multiple indefinite quantifiers. My experience has been that regular expressions with more than 2-3 `*` and/or `+` quantifiers tend to be hopelessly brittle and prone to weird results. Eg., are you ABSOLUTELY sure that you want to be always using greedy quantifiers (`*` and `+`) instead of reluctant quanitifiers (`*?` and `+?`) in your `maybeSome()` and `some()` methods? Have you thought about how these interact in every scenario? I can almost guarantee that there is at least one case where a `*` or `+` will eat too much of the cigar and lead to some erroneous result...

I feel that a straightforward parsing of the Cigar elements would be far less error-prone.